### PR TITLE
PDF part 4: Linearization, Parser error propagation, and color spaces

### PR DESCRIPTION
**Pull request by mattco98**
#### Parser

- Support linearized PDF documents. Most PDF files from the web will be linearized as it is more space efficient, so this is a big step in actually displaying PDFs straight from the web.
- All functions now propagate errors instead of crashing. This removed many (but not all) of the `VERIFY` calls in the parser. 

#### Renderer

- Fixed text rendering below the font baseline instead of above it
- Create `ColorSpace` class for more complex color spaces, and "implement" the `CalRGB` color space.
  - "implement" is in quotes because the PDF I have which requires this color space only uses it for black, which is trivial in terms of the color calculations. Eventually I'll have to manually craft a PDF which tests this further. 
- Avoid calculating ~4 matrix multiplications for every single text glyph, and instead only do them once when painting a string

#### PDF Viewer

- Switch icons back to up/down. I forgot to commit the image files for these in my last PR, and Linus changed them to left/right arrows. 
- Utilize the parser error propagation to display an error message to the user when parsing isn't successful

#### Misc

- Add `ByteBuffer::append(ReadonlyBytes)`
- Switch to east const (sorry Ali :sweat_smile:)

I was going to try to fix the text positioning in this PR as well, but I figured it was large enough. After fixing the text positioning, I will be pretty happy with where this little project is. From there it will just be about adding misc features like images and whatnot.

### CODE REVIEW

**Comment by alimpfard** [comment]
> Switch to east const (sorry Ali :sweat_smile:)

In the end, I'm forced to accept the project rules anyway :P

File: Userland/Libraries/LibPDF/Document.cpp [hunk 1]
```diff
@@ -34,20 +34,32 @@ String OutlineItem::to_string(int indent) const
...
-    : m_parser(Parser({}, bytes))
+RefPtr<Document> Document::create(const ReadonlyBytes& bytes)
 {
-    m_parser.set_document(this);
+    auto parser = adopt_ref(*new Parser({}, bytes));
+    auto document = adopt_ref(*new Document(parser));
 
-    VERIFY(m_parser.perform_validation());
-    auto [xref_table, trailer] = m_parser.parse_last_xref_table_and_trailer();
+    VERIFY(parser->perform_validation());
```
Comment (alimpfard) [inline Userland/Libraries/LibPDF/Document.cpp]: Why not just return from here?

File: Userland/Libraries/LibPDF/Document.cpp [hunk 2]
```diff
@@ -34,20 +34,32 @@ String OutlineItem::to_string(int indent) const
...
+    auto xref_table_and_trailer_opt = parser->parse_last_xref_table_and_trailer();
+    if (!xref_table_and_trailer_opt.has_value())
+        return {};
 
-    m_xref_table = xref_table;
-    m_trailer = trailer;
+    auto [xref_table, trailer] = xref_table_and_trailer_opt.value();
```
Comment (alimpfard) [inline Userland/Libraries/LibPDF/Document.cpp]:
```suggestion
    auto& [xref_table, trailer] = xref_table_and_trailer_opt.value();
```
or
```suggestion
    auto [xref_table, trailer] = xref_table_and_trailer_opt.release_value();
```

File: Userland/Libraries/LibPDF/Document.cpp [hunk 3]
```diff
@@ -34,20 +34,32 @@ String OutlineItem::to_string(int indent) const
...
 
-    m_catalog = m_trailer->get_dict(this, CommonNames::Root);
-    build_page_tree();
-    build_outline();
+    document->m_xref_table = xref_table;
+    document->m_trailer = trailer;
```
Comment (alimpfard) [inline Userland/Libraries/LibPDF/Document.cpp]: Please move these instead, they're not exactly cheap to copy : )

File: Userland/Applications/PDFViewer/PDFViewerWidget.cpp [hunk 4]
```diff
@@ -103,11 +104,18 @@ void PDFViewerWidget::open_file(const String& path)
...
+        GUI::MessageBox::show_error(nullptr, String::formatted("Couldn't open file: {}\n", path));
+        return;
+    }
+
     m_buffer = file_result.value()->read_all();
     auto document = PDF::Document::create(m_buffer);
-    // FIXME: Show error dialog if the Document is invalid
-    VERIFY(document);
+    if (!document) {
+        GUI::MessageBox::show_error(nullptr, String::formatted("Couldn't load PDF: {}\n", path));
```
Comment (alimpfard) [inline Userland/Applications/PDFViewer/PDFViewerWidget.cpp:115]: Is the newline intentional? seems a bit weird to have it in a message box

**Review by alimpfard — COMMENTED** [review]
Here's a very incomplete review (up to 07df34f5b5c2c148bc213f5416c7e85acaa2b9bb) :P

File: Userland/Libraries/LibPDF/Document.cpp (hunk 2 above)
Comment (mattco98) [reply Userland/Libraries/LibPDF/Document.cpp]: This is also changed later, and the table becomes a `RefPtr`

File: Userland/Libraries/LibPDF/Document.cpp (hunk 3 above)
Comment (mattco98) [reply Userland/Libraries/LibPDF/Document.cpp]: This is also changed later, and the table becomes a `RefPtr`

File: Userland/Applications/PDFViewer/PDFViewerWidget.cpp (hunk 4 above)
Comment (mattco98) [reply Userland/Applications/PDFViewer/PDFViewerWidget.cpp:115]: Uhh idk, I think I just copied it from some other code. I'll remove it though

File: Userland/Libraries/LibPDF/Renderer.h [hunk 5]
```diff
@@ -93,6 +93,7 @@ class Renderer {
     void handle_text_next_line_show_string_set_spacing(const Vector<Value>& args);
 
     // shift is the manual advance given in the TJ command array
```
Comment (FireFox317) [inline Userland/Libraries/LibPDF/Renderer.h:97]: This comment seems to refer to the `show_text` function, but is now refering to `set_graphics_state_from_dict`.

Comment (mattco98) [reply Userland/Libraries/LibPDF/Renderer.h]: Nice catch! Would have been confusing for anyone coming across that comment :)

**Pull request merged**